crates/schemalate/src/firebolt/firebolt_schema_builder.rs

```diff
@@ -0,0 +1,412 @@
+        .clone()
+        .ok_or(Error::FieldSelectionMissing)?;
+    let projections = binding
+        .collection
+        .clone()
+        .ok_or(Error::CollectionMissing)?
+        .projections;
+
+    let doc_field = if fs.document.len() > 0 {
+        vec![fs.document]
+    } else {
+        vec![]
+    };
+    let keys = fs.keys.clone();
+    let fields: Vec<String> = vec![fs.keys, fs.values, doc_field].concat();
+
+    let mut columns = Vec::new();
+
+    let errors = fields
+        .iter()
+        .map(|field| {
+            let projection = projections.iter().find(|p| &p.field == field).unwrap();
+            let inference = projection.inference.as_ref().unwrap();
+            let is_key = keys.contains(field);
+            let r#type = inference
+                .types
+                .iter()
+                .filter(|t| t != &"null")
+                .next()
+                .unwrap();
+
+            let fb_type = projection_type_to_firebolt_type(r#type).ok_or(Error::UnknownType {
+                r#type: r#type.to_string(),
+                field: field.clone(),
+            })?;
+
+            columns.push(Column {
+                key: projection.field.clone(),
+                r#type: fb_type,
+                nullable: inference.exists != i32::from(Exists::Must)
+                    || inference.types.contains(&"null".to_string()),
+                is_key,
+            });
+            Ok(())
+        })
+        .filter(|r| r.is_err())
+        .next();
+
+    if let Some(err) = errors {
+        // We have filtered on is_err, so this is a safe unwrap
+        return Err(err.unwrap_err());
+    }
+
+    Ok(TableSchema { columns })
+}
+
+pub fn build_firebolt_queries_bundle(
+    spec: MaterializationSpec,
+) -> Result<FireboltQueriesBundle, Error> {
+    let config: EndpointConfig = serde_json::from_str(&spec.endpoint_spec_json)?;
+
+    let bindings : Result<Vec<BindingBundle>, Error> = spec.bindings.iter().map(|binding| {
+        let resource: Resource = serde_json::from_str(&binding.resource_spec_json)?;
+        let mut schema = build_firebolt_schema(binding)?;
+
+        let external_table_name = format!("{}_external", resource.table);
+        let external_table = Table {
+            name: external_table_name.clone(),
+            r#type: TableType::External,
+            schema: schema.clone(),
+        };
+
+        // Add source_file_name column to main table
+        schema.columns.push(Column {
+            key: "source_file_name".to_string(),
+            r#type: FireboltType::Text,
+            is_key: false,
+            nullable: false,
+        });
+
+        let table = Table {
+            name: resource.table.clone(),
+            r#type: resource.table_type.into(),
+            schema: schema.clone(),
+        };
+
+        Ok(BindingBundle {
+            binding: 0, // FIXME
+            create_table: CreateTable {
+                table: table.clone(),
+                if_not_exists: true,
```

Comment:
Yeah I have the function there, it's commented out and `todo!()`! will work on that 👍🏽

Reply:
@jgraettinger I was thinking about this, and if I understand correctly, Postgres doesn't do this: With materialize-postgres, if the table is not already created, Flow creates it, otherwise we re-use the given table (that will only work if the table actually conforms to the schema, of course). Is there a reason we might want to force creation of table unless we ourselves have created the table? (This is a separate point from validating the fields)

Reply:
`materialize-postgres` does issue a `CREATE TABLE` (without `IF NOT EXISTS`), but it only issues that CREATE TABLE if the applied spec introduces a newly bound table resource that wasn't present in a previously-applied spec.

I think it's fair to question whether the connector should attempt to use an already-extant table. The rationale for _not_ using `IF NOT EXISTS` is that it prevents confusion / it's very explicit to the user what the table state is.

This is probably a broader conversation than what we should have in a PR though. If you'd prefer to leave the current handling for now, sounds good to me 👍

Reply:
Hmm, I saw this: https://github.com/estuary/flow/pull/414/files#diff-e1407d4391856ac4d711d56cc9696097248d8e53800ced6b7044ef253e764634R12

That's why I thought materialize-postgres allows for an existing table to be provided. If not, then let's make sure we reflect it in the docs at least

Reply:
Good catch! I questioned my own recollection given this, but after [double-checking](https://github.com/estuary/flow/blob/5270e9d3612815edac9c8673420c91215260aa3b/go/protocols/materialize/sql/spec_mapping.go#L13-L20) did confirm that it's _not_ using `IF NOT EXISTS` today (if it were, `&Table{IfNotExists: true}` would be set).